### PR TITLE
✨ analytics: Allow pings with totalVisibleTime=0 when spec has reportWhen

### PR DESCRIPTION
**Pull request by warrengm**
This change allows pings to be sent by amp-analytics when `visibleTime` is 0 provided:

1. The spec explicitly contains `totalTimeMin: 0`  and `continuousTimeMin: 0` to verify that it's intended behavior by the user. Otherwise the implicit behavior is to treat the lower bound as exclusive so no ping will be sent.

2. The spec contains `reportWhen` which causes the ping to be defered to some future event. Otherwise the ping would be sent immediately with any chance to collect useful information.

In short, this PR enables the following config:

```
"visibilitySpec": {
  "reportWhen": "documentExit",
  "visiblePercentageMin": 50,  // Visibility threshold
  "totalTimeMin": 0,
  "continuousTimeMin": 0
},
"extraUrlParams": {
  "visTime": "${totalVisibleTime}"  // Time that >50% of element was on screen, may be 0ms
}
```

### CODE REVIEW

**Comment by zhouyx** [comment]
@warrengm Have one question about the PR. If I understand it correctly, this PR attempts to introduce a way to trigger a visibility ping no matter what. 
What do you think about having `visiblePercentageMin: 0 & visibilityPercentageMax: 0` instead of 
`totalTimeMin: 0 and continuousTimeMin: 0`? Also I wonder if `visiblePercentageMin: 0 & visibilityPercentageMax: 0` already work.

**Comment by warrengm** [comment]
@zhouyx That's right. We want to get a ping even if `totalVisibleTime` is zero since there are some other parameters we care about.

I tested `visiblePercentageMin: 0` and `visibilityPercentageMax: 0` and it currently sends a ping right away without waiting for `reportWhen`. This PR fixes this by changing the default value of `reportReady_`.

However, with the above, `${totalVisibleTime}` provides the time when the element was _not_ on screen. I'd like a way to get the total time when 50% of the element was on screen. I could imagine having a variable like `${totalVisibleTime(50)}` but I think such a change adds more complexity than what's needed. I'd be happy to change this PR if you have a better suggestion.

**Comment by zhouyx** [comment]
> I tested visiblePercentageMin: 0 and visibilityPercentageMax: 0 and it currently sends a ping right away without waiting for reportWhen. This PR fixes this by changing the default value of reportReady_.

This sounds like a bug to me that we should fix. 

>However, with the above, ${totalVisibleTime} provides the time when the element was not on screen. I'd like a way to get the total time when 50% of the element was on screen.

I see. So the you will need to set `visibilePercentageMin: 50` to calculate the correct totalVisibleTime. However by setting that, the request won't be sent out if the element hasn't been viewed. 

To force sending out a visibility request. I propose we 

Introduce a 'forceSend' flag to visibilitySpec
Or 
Instead of setting `totalTimeMin` & `continousTimeMin` to 0, but setting `totalTimeMax` to 0 instead. (As in theory that should not happen) 

cc @jeffjose

**Comment by warrengm** [comment]
The new flag approach sounds good. I'll restrict the flag's semantics to when `reportWhen` is present. I'll also update documentation.

**Comment by warrengm** [comment]
Done with implementation, testing, and documentation. All presubmits have passed

File: extensions/amp-analytics/0.1/visibility-model.js [hunk 1]
```diff
@@ -51,6 +51,17 @@ export class VisibilityModel {
...
 
+    /** @private @const {boolean} */
+    this.accumulateCounters_ = spec['reportWhen'] !== undefined;
+
+    /**
+     * Allow zero visible time on screen when explicitly provided and when the
+     * ping is deferred to some other event.
+     * @private @const {boolean}
+     */
+    this.forceReport_ =
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/visibility-model.js]: We need to make sure `hidden` trigger, `reportWhen`, `forceReport` and `repeat` work well together. 
Right now: `hidden` trigger equals to `reportWhen: pageHidden`. `reportWhen` doesn't work with `repeat` with the special case, `hidden` trigger works with `repeat`. 

We can be more strict on checking `forceReport` only works with `reportWhen` but not `repeat: true`. One approach is to have `reportWhen` be an object and have `forceReport` be one of its field. But there's the tradeoff making the configuration too complicated. 

WDYT? cc @lannka

File: extensions/amp-analytics/0.1/visibility-model.js [hunk 2]
```diff
@@ -298,7 +309,7 @@ export class VisibilityModel {
       return;
     }
     const conditionsMet = this.updateCounters_(visibility);
-    if (conditionsMet) {
+    if (conditionsMet || this.forceReport_) {
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/visibility-model.js]: nit: could we have `conditionsMet = this.updateCounters_(visibility) || this.forceReport_;` instead.  And comment please

File: extensions/amp-analytics/amp-analytics.md [hunk 3]
```diff
@@ -442,7 +442,7 @@ NOTE: There is a [known issue](https://github.com/ampproject/amphtml/issues/1089
...
   - `waitFor`: This property indicates that the visibility trigger should wait for a certain signal before tracking visibility. The supported values are `none`, `ini-load` and `render-start`. If `waitFor` is undefined, it is defaulted to [`ini-load`](#initial-load-trigger) when selector is specified, or to `none` otherwise.
-  - `reportWhen`: This property indicates that the visibility trigger should wait for a certain signal before sending the trigger. The only supported value is `documentExit`. `reportWhen` and `repeat` may not both be used in the same visibilitySpec.
+  - `reportWhen`: This property indicates that the visibility trigger should wait for a certain signal before sending the trigger. The only supported value is `documentExit`. `reportWhen` and `repeat` may not both be used in the same visibilitySpec. When `reportWhen` is specified, the `visibilitySpec` may include the `forceReport` property to indicate that the report should be sent at the time of the signal even if visibility requirements have not been met. In that case, any relevant variables (`totalVisibleTime`, etc.) will default to 0.
```
Comment (zhouyx) [inline extensions/amp-analytics/amp-analytics.md]: Question: I don't believe the `totalVisibleTime` will default to 0 with `forceReport`. Thought the idea of introducing `forceReport` was to collect the totalVisibleTime above certain percentage.

File: extensions/amp-analytics/0.1/test/test-visibility-model.js [hunk 4]
```diff
@@ -1121,4 +1122,111 @@ describes.sandboxed('VisibilityModel', {}, () => {
...
   });
+
+  describe('reportWhen', () => {
+    beforeEach(() => {
+      sandbox.restore(); // Restore timers.
+    });
+
+    // Yields the thread to allow internal promises to complete.
+    const yieldThread =
+      async() => new Promise(resolve => setTimeout(resolve, 1));
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/test/test-visibility-model.js]: nit: would `setTimeout(,0)` work? You could do `yield macroTask()` for that.

File: extensions/amp-analytics/0.1/test/test-visibility-model.js [hunk 5]
```diff
@@ -1121,4 +1122,111 @@ describes.sandboxed('VisibilityModel', {}, () => {
       expect(spy).to.be.calledTwice;
     });
   });
+
+  describe('reportWhen', () => {
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/test/test-visibility-model.js]: Could you leverage the existed tests for `reportReadyPromise` https://github.com/ampproject/amphtml/blob/master/extensions/amp-analytics/0.1/test/test-visibility-model.js#L455
Thanks.

File: extensions/amp-analytics/0.1/visibility-model.js [hunk 6]
```diff
@@ -79,7 +90,7 @@ export class VisibilityModel {
     this.ready_ = true;
 
     /** @private {boolean} */
-    this.reportReady_ = true;
+    this.reportReady_ = !this.accumulateCounters_;
```
[the remark this answers is not in the document]
Comment (warrengm) [reply extensions/amp-analytics/0.1/visibility-model.js]: This fixes the race condition where the ping is sent right away without waiting on the `reportReadyPromise` because the thread is yielded before `setReportReady` is called. So we'll end up sending a ping if `update` is called too early.

Added a comment clarifying. I'd be happy to cleanup the code in a follow-up PR. What if  `reportReady_` is unconditionally false and the visibility manager calls `setReportReady(true)` or `setReportReady(Promise.resolve())` when `reportWhen` is missing?

File: extensions/amp-analytics/0.1/visibility-model.js (hunk 1 above)
Comment (warrengm) [reply extensions/amp-analytics/0.1/visibility-model.js]: I'm not fully familiar with the behavior of `repeat`. Will it force a report?

In this case, it'd be possible to get force multiple pings if the `reportWhen` fires multiple times. (Which, to confirm, can happen with the `hidden` trigger as it's based on `visibilitychange` events?) This seems like possibly the right behavior, e.g. to implement a heartbeat ping. Though maybe we want to toggle `forceReport` on repeated triggers to prevent users from implementing heartbeat pings?

File: extensions/amp-analytics/amp-analytics.md (hunk 3 above)
Comment (warrengm) [reply extensions/amp-analytics/amp-analytics.md]: It defaults to 0 in local testing.

The goal of `forceReport` is to be able to report `totalVisibleTime` even when its zero. We usually set `visiblePercentageMin` to 50% but we want to send pings (with `totalVisibleTime=0`) when zero pixels were on screen, 40% of pixels were on screen, etc.

File: extensions/amp-analytics/0.1/test/test-visibility-model.js (hunk 4 above)
Comment (warrengm) [reply extensions/amp-analytics/0.1/test/test-visibility-model.js]: Thanks for the reference!

File: extensions/amp-analytics/0.1/test/test-visibility-model.js (hunk 5 above)
Comment (warrengm) [reply extensions/amp-analytics/0.1/test/test-visibility-model.js]: Done but I couldn't share much with those tests because:

1. The spec changes between tests
2. I avoided `eventSpy = vh.eventResolver_.resolver` to avoid relying on private variables. I use `onTriggerEvent(eventSpy)` to verify that it's reports are sent only once. Using `eventResolver_` seems fragile to me since the `eventResolver_` is re-instantiated in `reset_` but the event spy is not updated.

File: extensions/amp-analytics/0.1/test/test-visibility-model.js (hunk 4 above)
Comment (warrengm) [reply extensions/amp-analytics/0.1/test/test-visibility-model.js]: No longer applicable

File: extensions/amp-analytics/0.1/visibility-model.js (hunk 2 above)
Comment (warrengm) [reply extensions/amp-analytics/0.1/visibility-model.js]: Done

**Comment by warrengm** [comment]
Made changes. PTAL

**Comment by zhouyx** [comment]
@lannka Please take a look at the config design. 
As we talked offline, `forceReport` is what I come up to support the use case here. And there's tradeoff between making it a standalone field of a field of `reportWhen`.

**Comment by zhouyx** [comment]
@warrengm I talked to @lannka offline. @lannka argued that `reportWhen` should be associated with `forceReport` out of the box. 

Does your use case, or do you come up with other use case that a spec with `reportWhen` should only be triggered after visibilitySpec has been met? (Which @lannka 's proposal doesn't work?) 

If not, we could instead making the `forceReport` an internal field now. 

@lannka Please add your thoughts that I fail to summarize here.

**Comment by warrengm** [comment]
@zhouyx I can't think of any cases where `reportWhen` requires `visibilitySpec` to have been met. The only case I can think if is batching repeated triggers onto the same ping--but that not supported by amp-analytics, right?

If I understand correctly, you're recommending removing an explicit `forceReport` from the config and having that be the default behavior of `reportWhen`, right?

**Comment by lannka** [comment]
@zhouyx that sounds right. To share more thinking behind the syntax:

`reportWhen` describes what the triggering condition is, it has 3 possible values:
- visible (default)
- documentHidden
- documentExit
 
**1) visible**
`reportWhen: visible` means we will check if all the "visibility conditions" are met, including
```
totalVisibleTimeMin/Max (default 0/inf)
continuousVisibleTimeMin/Max (default 0/inf) 
```

where the visible time are calculated based on element's "visibility status", which is defined by
```
selector
visiblePercentageMin/Max (default 0/100)
```
Note that this "visibility status" is also used to calculate metrics like `totalVisibleTime`.

**2) documentHidden / documentExit**
For `reportWhen: documentHidden / documentExit`, we will always trigger upon on the corresponding events, ignoring the "visibility conditions".

Note that for legacy reason, we have a separate "hidden" trigger, which right now shares code with "reportWhen: documentHidden". To not affect existing users, we might start to separate the logic.

**Comment by warrengm** [comment]
Sounds good. I'll make the changes for (2) and have a special case for `on: hidden` to prevent forced reports.

My understanding is that (1) is not an issue because the `reportWhen: visible` syntax is not currently supported per [events.js](https://github.com/ampproject/amphtml/blob/36117c4d00b888dd12323e021861a3b10818f143/extensions/amp-analytics/0.1/events.js#L1170) and [user documentation](https://www.ampproject.org/docs/reference/components/amp-analytics). Users would simply use `on: visible`.

**Comment by warrengm** [comment]
Updated the implementation. PTAL

I changed the `reportReady_` default so that it's always false and moved the bug fix into the visibility manager. As a follow up PR, I'd like to make the default values to false to avoid introducing the bug again in the future.

File: extensions/amp-analytics/amp-analytics.md (hunk 3 above)
Comment (zhouyx) [reply extensions/amp-analytics/amp-analytics.md]: Right, but it doesn't default to 0. I believe it should still be the time that's between the specified visibility percentage. 

When you say it defaults to 0 in local test? What is the expected `totalVisibleTime`? Is it true that the element hasn't be 50% visible before?

File: extensions/amp-analytics/0.1/visibility-model.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (zhouyx) [reply extensions/amp-analytics/0.1/visibility-model.js]: I agree. Let's address this in following PR, to always have visibility manager setReportReady. Also this could avoid having to check `spec['waitFor']` again in visibility model.

File: extensions/amp-analytics/0.1/visibility-model.js (hunk 1 above)
Comment (zhouyx) [reply extensions/amp-analytics/0.1/visibility-model.js]: Not an issue any more given we are going to change the `reportWhen` default behavior.

File: extensions/amp-analytics/0.1/test/test-visibility-model.js [hunk 7]
```diff
@@ -24,6 +25,13 @@ describes.sandboxed('VisibilityModel', {}, () => {
   let startTime;
   let clock;
 
+  const tick = async(timeout = 0) => {
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/test/test-visibility-model.js]: ditto: does `yield macroTask()` work here?

File: extensions/amp-analytics/0.1/test/test-visibility-model.js [hunk 8]
```diff
@@ -164,8 +172,7 @@ describes.sandboxed('VisibilityModel', {}, () => {
     });
   });
 
-  describe('structure', () => {
-    let visibility;
+  describe('structure', () => { let visibility;
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/test/test-visibility-model.js]: nit: can we revert the change here

File: extensions/amp-analytics/0.1/test/test-visibility-model.js [hunk 9]
```diff
@@ -456,37 +463,75 @@ describes.sandboxed('VisibilityModel', {}, () => {
       let reportPromise;
       let promiseResolver;
       beforeEach(() => {
-        reportPromise = new Promise(resolve => {
-          promiseResolver = resolve;
-        });
-        vh.setReportReady(() => {return reportPromise;});
+        promiseResolver = new Deferred();
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/test/test-visibility-model.js]: nit: naming, is it possible to either rename `promiseResolver`, or make it the resolver function?

File: extensions/amp-analytics/0.1/test/test-visibility-model.js [hunk 10]
```diff
@@ -456,37 +463,75 @@ describes.sandboxed('VisibilityModel', {}, () => {
...
+        });
+
+        const shouldTriggerEventTestSpecs = [
+          {reportWhen: 'documentExit'},
+          {reportWhen: 'documentExit', totalTimeMin: 100000,
+            visiblePercentageMin: 50},
+        ];
+
+        for (const i in shouldTriggerEventTestSpecs) {
+          it('should trigger event with forceReport,' +
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/test/test-visibility-model.js]: nit: with reportWhen = documentExit

File: extensions/amp-analytics/0.1/test/test-visibility-model.js [hunk 11]
```diff
@@ -456,37 +463,75 @@ describes.sandboxed('VisibilityModel', {}, () => {
...
         });
       });
+
+      describe('with reportWhen', () => {
+        beforeEach(() => {
+          visibilityValueForTesting = 0;
+        });
+
+        const shouldTriggerEventTestSpecs = [
+          {reportWhen: 'documentExit'},
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/test/test-visibility-model.js]: Can we also add test to `reportWhen = 'documentHidden'`?

File: extensions/amp-analytics/0.1/test/test-visibility-manager.js [hunk 12]
```diff
@@ -1082,6 +1082,75 @@ describes.realWin('VisibilityManager integrated', {amp: true}, env => {
     });
   });
 
+  it('should wait for readyReportPromise', async() => {
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/test/test-visibility-manager.js]: These tests lives under the VisibilityManager integrated test category, but they don't looks like integrated test to me. 

We can keep these test, but can we add more integrated test that test the resolved state value.

I am imaging two tests
1. with `reportWhen`, and `visiblePercentageMin`.  visible percentage never meet, check the model resolved with correct state. 
2. with `reportWhen`, `visiblePercentageMin`, `continousTimeMin`, continous time never meet, check the model resolved with correct state. 

I feel 2 could probably replace 1. Let me know if you think there's better ways to test the end to end logic here. Thanks.

File: extensions/amp-analytics/0.1/test/test-visibility-model.js (hunk 7 above)
Comment (warrengm) [reply extensions/amp-analytics/0.1/test/test-visibility-model.js]: It would, but I imagine that this could be used in other tests that accumulate time while avoiding testing private fields. Let me know what you think

File: extensions/amp-analytics/0.1/visibility-model.js [hunk 13]
```diff
@@ -51,6 +51,13 @@ export class VisibilityModel {
       this.spec_['visiblePercentageMax'] = 0;
     }
 
+    /**
+     * Accumulate visibility counters but do not fire the trigger until the
+     * ready promise resolves.
+     * @private @const {boolean}
+     */
+    this.accumulateCounters_ = spec['reportWhen'] !== undefined;
```
[the remark this answers is not in the document]
Comment (warrengm) [reply extensions/amp-analytics/0.1/visibility-model.js]: `on: hidden` doesn't mutate the spec JSON so it should WAI: https://github.com/ampproject/amphtml/blob/master/extensions/amp-analytics/0.1/events.js#L1196. I'll add an E2E test for this

I'll go with `ignoreCountersForReport_`

File: extensions/amp-analytics/0.1/test/test-visibility-manager.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (warrengm) [reply extensions/amp-analytics/0.1/test/test-visibility-manager.js]: I've always used async functions in tests. Will switch to generators.

File: extensions/amp-analytics/0.1/test/test-visibility-model.js (hunk 7 above)
Comment (warrengm) [reply extensions/amp-analytics/0.1/test/test-visibility-model.js]: Actually, I think it's important that we don't yield the microtask queue when handling the `unload` event fires so I'll stick to `Promise.resolve()` and leave a comment. (I actually haven't tested it but It's my guess. We should add an E2E test for this)

File: extensions/amp-analytics/0.1/test/test-visibility-manager.js (hunk 12 above)
Comment (warrengm) [reply extensions/amp-analytics/0.1/test/test-visibility-manager.js]: Done. PTAL

File: extensions/amp-analytics/0.1/visibility-model.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (zhouyx) [reply extensions/amp-analytics/0.1/visibility-model.js]: Sounds good. Thanks.

File: extensions/amp-analytics/0.1/test/test-visibility-manager.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (zhouyx) [reply extensions/amp-analytics/0.1/test/test-visibility-manager.js]: Yes, it's completely a personal preference. But I've checked our codebase, and found we tend to prefer generator : ) Thanks.

File: extensions/amp-analytics/0.1/test/test-visibility-model.js (hunk 7 above)
Comment (zhouyx) [reply extensions/amp-analytics/0.1/test/test-visibility-model.js]: Sorry I didn't fully understand the reason why we can't yield the microTask queue here. 

The async/await and the yield with generators really look the same to me here. Could you further explain it? Thank you.

File: extensions/amp-analytics/0.1/test/test-visibility-manager.js [hunk 14]
```diff
@@ -1082,59 +1082,237 @@ describes.realWin('VisibilityManager integrated', {amp: true}, env => {
...
+      backgroundedAtStart: 0,
+      elementHeight: 100,
+      elementWidth: 100,
+      elementX: 0,
+      elementY: 75,
+      firstSeenTime: 50,
+      lastSeenTime: 90,
+      lastVisibleTime: 0, // Didn't meet visibility.
+      loadTimeVisibility: 25,
+      maxVisiblePercentage: 0,
```
[the remark this answers is not in the document]
Comment (warrengm) [reply extensions/amp-analytics/0.1/test/test-visibility-manager.js]: I think that might be related to `maxContinuousVisibleTime` being 0. I'll confirm and file an issue + add a comment here if so.

File: extensions/amp-analytics/0.1/test/test-visibility-manager.js [hunk 15]
```diff
@@ -1082,59 +1082,237 @@ describes.realWin('VisibilityManager integrated', {amp: true}, env => {
...
+          lastVisibleTime: 50,
+          loadTimeVisibility: 25,
+          maxVisiblePercentage: 25,
+          minVisiblePercentage: 25,
+          totalVisibleTime: 30,
+          maxContinuousVisibleTime: 30,
+        });
+      });
+
+  it('should wait for readyPromise with reportReadyPromise', async() => {
```
[the remark this answers is not in the document]
Comment (warrengm) [reply extensions/amp-analytics/0.1/test/test-visibility-manager.js]: Ah, oops. Done

File: extensions/amp-analytics/0.1/test/test-visibility-manager.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (warrengm) [reply extensions/amp-analytics/0.1/test/test-visibility-manager.js]: Verified. Adding another `fireIntersect` after the ready resolver didn't help. Will file an issue

**Comment by warrengm** [comment]
@zhouyx I can't see the diffs to look into the increase in bundle size. Could it be due to the fact that my branch is behind HEAD?

**Comment by zhouyx** [comment]
@warrengm Not sure, the change shouldn't increase bundle size. Rebasing might help.

**Comment by rsimha** [comment]
From what I can tell, this PR shouldn't have increased the bundle size, so I wonder if there's an error in how it was computed. @danielrozenberg should be able to tell you why this happened, and how to approve the build.

**Comment by danielrozenberg** [comment]
Oh, I see that the Travis check ran 4 days ago, before all the little changes we just introduced. I'll rerun the Travis check which should hopefully merge correctly with master

**Comment by zhouyx** [comment]
oops, realized the check is non blocking and merged already : )

**Comment by danielrozenberg** [comment]
Welp, once it's merged the Travis checks fail. Fingers crossed that we haven't broken everything *this-is-fine-dog.gif*

**Comment by warrengm** [comment]
Thanks for the review!
